File: code/ex1096.java
class Solution {
    public int i = 0, n;
    
    public List<String> braceExpansionII(String S) {
        n = S.length();
        Set<String> ret = dfs(S);
        List<String> res = new ArrayList<>(ret);
        Collections.sort(res);
        return res;
    }
    
    private Set<String> dfs(String S) {
        Set<String> ret = new HashSet<>();
        Set<String> temp = new HashSet<>();
        String s = "";
        while (i < S.length()) {
            char c = S.charAt(i);
            if (c >= 'a' && c <= 'z') {    // Is alphabet
                s += c;
                i++;
            }
            else {
                if (!s.equals("")) {
                    if (temp.isEmpty()) {
                        temp.add(s);
                        s = "";
                    }
                    else {
                        temp = addString(temp, s);
                        s = "";
                    }
                }
                
                if (c == '{') {
                    i++;
                    Set<String> next = dfs(S);
                    temp = combine(temp, next);
                }
                else if (c == '}') {
                    if (!temp.isEmpty()) {
                        addList(ret, temp);
                        temp.clear();
                    }
                    i++;
                    return ret;
                }
                else if (c == ',') {
                    if (!temp.isEmpty()) {
                        addList(ret, temp);
                        temp.clear();
                    }
                    i++;
                }
            }
        }
        
        if (!s.equals("")) {
            if (temp.isEmpty())    temp.add(s);
            else    temp = addString(temp, s);
        }
        
        if (!temp.isEmpty())    addList(ret, temp);
        
        return ret;
    }
    
    private Set<String> combine(Set<String> l1, Set<String> l2) {
        if (l1.isEmpty())    return l2;
        Set<String> combined = new HashSet<>();
        for (String s1 : l1) {
            for (String s2 : l2) {
                combined.add(s1 + s2);
            }
        }
        return combined;
    }
    
    private void addList(Set<String> res, Set<String> temp) {
        for (String t : temp)
            res.add(t);
    }
    
    private Set<String> addString(Set<String> set, String s) {
        Set<String> res = new HashSet<>();
        for (String se : set) {
            res.add(se + s);
        }
        return res;
    }
}

File: code/ex1096.cpp
class Solution {
public:
    vector<string> braceExpansionII(string expression) {
        // Wrap alphabet characters with {}
        string s = "";
        for (char c : expression) {
            if (isalpha(c)) {
                s = s + '{' + c + '}';
            } else {
                s += c;
            }
        }
        
        // Represents
        stack<unordered_set<string>> stk;
        stack<int> oper;    // 0 -> ',' before, 1 -> '{' before
        unordered_set<string> temp;
        
        for (char c : s) {
            if (c == '{') {
                stk.push(temp);
                temp.clear();
                oper.push(1);
            } else if (c == ',') {
                stk.push(temp);
                temp.clear();
                oper.push(0);
            } else if (c == '}') {
                while (oper.top() == 0) {
                    temp = combine(stk.top(), temp);
                    stk.pop();
                    oper.pop();
                } 
                if (oper.top() == 1) {
                    temp = concat(stk.top(), temp);
                    stk.pop();
                    oper.pop();
                }
            } else {
                string ss = "";
                ss += c;
                temp.insert(ss);
            }
        }
        
        // Get result
        vector<string> res(temp.begin(), temp.end());
        sort(res.begin(), res.end());
        return res;
    }
    
private:
    unordered_set<string> combine(unordered_set<string>& s1, unordered_set<string>& s2) {
        unordered_set<string> res;
        for (string s : s1)    res.insert(s);
        for (string s : s2)    res.insert(s);
        return res;
    }
    
    unordered_set<string> concat(unordered_set<string>& s1, unordered_set<string>& s2) {
        if (s1.size() == 0)    s1.insert("");
        unordered_set<string> res;
        for (string s : s1) {
            for (string t : s2) {
                res.insert(s + t);
            }
        }
        return res;
    }
};